File: vuln-sandbox/python/cwe-532/password-in-log-safe.py
# SAFE: password-in-log — password field excluded from log output
# Rule: AuthPasswordInLog | CWE-532 | Expected: TrueNegative

import logging
from flask import Flask, request, jsonify

app = Flask(__name__)
logger = logging.getLogger(__name__)

SENSITIVE_FIELDS = {'password', 'token', 'secret', 'api_key', 'authorization'}


def sanitize_for_log(data: dict) -> dict:
    """Return a copy of data with sensitive fields redacted."""
    return {k: '[REDACTED]' if k.lower() in SENSITIVE_FIELDS else v for k, v in data.items()}


@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()

    # SAFE: log only the sanitized request; password is never written to logs
    logger.info('Login attempt: %s', sanitize_for_log(data))

    username = data.get('username')
    password = data.get('password')

    if username == 'admin' and password == 'correct-password':
        return jsonify({'token': 'jwt-token-here'})
    return jsonify({'error': 'Invalid credentials'}), 401


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/node/cwe-532/password-in-log-safe.js
// SAFE: password-in-log — password field excluded from log output
// Rule: AuthPasswordInLog | CWE-532 | Expected: TrueNegative

const express = require('express');
const app = express();
app.use(express.json());

function sanitizeForLog(obj) {
  const SENSITIVE_FIELDS = ['password', 'token', 'secret', 'apiKey', 'authorization'];
  const sanitized = { ...obj };
  for (const field of SENSITIVE_FIELDS) {
    if (field in sanitized) sanitized[field] = '[REDACTED]';
  }
  return sanitized;
}

app.post('/login', (req, res) => {
  const { username, password } = req.body;

  // SAFE: log only the sanitized request body; password is redacted
  console.log('Login attempt:', JSON.stringify(sanitizeForLog(req.body)));

  if (username === 'admin' && password === process.env.ADMIN_PASSWORD) {
    res.json({ token: 'jwt-token-here' });
  } else {
    res.status(401).json({ error: 'Invalid credentials' });
  }
});

app.listen(3000);
